packages/api/amplify_api/lib/src/graphql/ws/web_socket_connection.dart

```diff
@@ -0,0 +1,262 @@
+    _subscription?.cancel();
+    _channel?.sink.done.whenComplete(() => _channel = null);
+    _channel?.sink.close(closeStatus, reason);
+    _rebroadcastController.close();
+    _timeoutTimer?.cancel();
+    _resetConnectionInit();
+  }
+
+  /// Initializes the connection.
+  ///
+  /// Connects to WebSocket, sends connection message and resolves future once
+  /// connection_ack message received from server. If the connection was previously
+  /// established then will return previously completed future.
+  Future<void> init() => _initMemo.runOnce(_init);
+
+  Future<void> _init() async {
+    final connectionUri =
+        await generateConnectionUri(_config, _authProviderRepo);
+    await connect(connectionUri);
+
+    send(WebSocketConnectionInitMessage());
+
+    return ready;
+  }
+
+  Future<void> _sendSubscriptionRegistrationMessage<T>(
+      GraphQLRequest<T> request) async {
+    await init(); // no-op if already connected
+    final subscriptionRegistrationMessage =
+        await generateSubscriptionRegistrationMessage(
+      _config,
+      id: request.id,
+      authRepo: _authProviderRepo,
+      request: request,
+    );
+    send(subscriptionRegistrationMessage);
+  }
+
+  /// Subscribes to the given GraphQL request. Returns the subscription object,
+  /// or throws an [Exception] if there's an error.
+  Stream<GraphQLResponse<T>> subscribe<T>(
+    GraphQLRequest<T> request,
+    void Function()? onEstablished,
+  ) {
+    // Create controller for this subscription so we can add errors.
+    late StreamController<GraphQLResponse<T>> controller;
+    controller = StreamController<GraphQLResponse<T>>.broadcast(
+      onCancel: () {
+        _cancel(request.id);
+        controller.close();
+      },
+    );
+
+    // Filter incoming messages that have the subscription ID and return as new
+    // stream with messages converted to GraphQLResponse<T>.
+    _messageStream
+        .where((msg) => msg.id == request.id)
+        .transform(WebSocketSubscriptionStreamTransformer(
+          request,
+          onEstablished,
+          logger: _logger,
+        ))
+        .listen(
+          controller.add,
+          onError: controller.addError,
+          onDone: controller.close,
+          cancelOnError: true,
+        );
+
+    _sendSubscriptionRegistrationMessage(request)
+        .catchError(controller.addError);
+
+    return controller.stream;
+  }
+
+  /// Cancels a subscription.
+  void _cancel(String subscriptionId) {
+    _logger.info('Attempting to cancel Operation $subscriptionId');
+    send(WebSocketStopMessage(id: subscriptionId));
+    // TODO(equartey): if this is the only subscription, close the connection.
+  }
+
+  /// Serializes a message as JSON string and sends over WebSocket channel.
+  @visibleForTesting
+  void send(WebSocketMessage message) {
+    final msgJson = json.encode(message.toJson());
+    if (_channel == null) {
+      throw ApiException(
```

Comment:
agreed, channel shouldn't be null here so just changed  to use `!` here. Channel is only null during connection init so that shouldn't be a factor. I think this kind of logic might get touched more when implementing reconnect but will leave that for Elijah.